### PR TITLE
Further small HashTable optimizations

### PR DESCRIPTION
**Pull request by l0rinc**
These are rather refactorings, as I cannot measure any real difference with the existing `JMH` benchmarks.

### CODE REVIEW

File: src/library/scala/collection/mutable/HashTable.scala [hunk 1]
```diff
@@ -367,11 +367,7 @@ trait HashTable[A, Entry >: Null <: HashEntry[A, Entry]] extends HashTable.HashU
     * Note: we take the most significant bits of the hashcode, not the lower ones
     * this is of crucial importance when populating the table in parallel
     */
-  protected final def index(hcode: Int): Int = {
-    val ones = table.length - 1
-    val exponent = Integer.numberOfLeadingZeros(ones)
-    (improve(hcode, seedvalue) >>> exponent) & ones
-  }
+  protected final def index(hcode: Int): Int = if (table.length == 1) 0 else improve(hcode, seedvalue) >>> numberOfLeadingZeros(table.length - 1)
```
Comment (l0rinc) [inline src/library/scala/collection/mutable/HashTable.scala]: See: https://github.com/scala/scala/pull/5566#discussion_r92026160 and https://github.com/scala/scala/pull/5566#discussion_r92055219

File: src/library/scala/collection/mutable/HashTable.scala [hunk 2]
```diff
@@ -429,16 +429,7 @@ private[collection] object HashTable {
...
-    /* See http://bits.stephan-brumme.com/roundUpToNextPowerOfTwo.html */
-    var c = target - 1
-    c |= c >>>  1
-    c |= c >>>  2
-    c |= c >>>  4
-    c |= c >>>  8
-    c |= c >>> 16
-    c + 1
-  }
+  private[collection] def nextPositivePowerOfTwo(target : Int): Int = 1 << -numberOfLeadingZeros(target - 1)
```
Comment (l0rinc) [inline src/library/scala/collection/mutable/HashTable.scala]: should probably be extracted in a static method somewhere outside instead, it's not related strictly to maps, rather integers.

Comment (l0rinc) [reply src/library/scala/collection/mutable/HashTable.scala]: Related to https://github.com/scala/scala/pull/5569#discussion_r91168262

**Comment by l0rinc** [comment]
@szeiger, I get
```Found 2 binary incompatibilities (13 were filtered out)
=======================================================
 * the type hierarchy of object
   scala.concurrent.forkjoin.package#ThreadLocalRandom is different in current
   version. Missing types {scala.Serializable,java.io.Serializable}
 * the type hierarchy of object
   scala.concurrent.forkjoin.package#ForkJoinTask is different in current
   version. Missing types {scala.Serializable,java.io.Serializable}
```
but I didn't modify any of those. Should I add them to the exception list also?

**Comment by SethTisue** [comment]
The MiMa thing is unrelated. There's a recent scala-dev ticket on it

**Comment by l0rinc** [comment]
Thank you @SethTisue, will rebase it, once it's solved! :)

**Comment by l0rinc** [comment]
Rebased :)

**Comment by SethTisue** [comment]
#5626 (which replaced #5566) was merged onto 2.11.x. everything merged onto 2.11.x eventually gets merged forward onto 2.12.x and 2.13.x, too. so is this PR still needed...?

**Comment by l0rinc** [comment]
Hey @SethTisue, this contains an [additional commit](https://github.com/scala/scala/pull/5600/commits/9af88bb78fb8829ef43635ffccf15d01914b1106), please decide if you think it's worth keeping or not, I won't mind either way.
Thanks for your help :)

**Comment by SethTisue** [comment]
let's wait until after the next 2.11 -> 2.12 merge, then rebase this

**Comment by l0rinc** [comment]
@SethTisue, removed the duplicate commit, thanks.

**Comment by l0rinc** [comment]
Failure seems unrelated, again:
```
ERROR: Failed to run DSL Script
groovy.lang.MissingPropertyException: No such property: Result for class: Script1
```

**Comment by retronym** [comment]
The root error can be found in the logs of the [`validate-publish-core`](https://scala-ci.typesafe.com/job/scala-2.12.x-validate-publish-core/4528/console) job:

```
[info] Running scala.tools.docutil.ManMaker fsc, scala, scalac, scaladoc, scalap /home/jenkins/workspace/scala-2.12.x-validate-publish-core/target/scala-dist/resource_managed/main/doc/tools /home/jenkins/workspace/scala-2.12.x-validate-publish-core/target/scala-dist/resource_managed/main/genman
[error] /home/jenkins/workspace/scala-2.12.x-validate-publish-core/src/library/scala/collection/mutable/HashTable.scala:436: not found: value numberOfLeadingZeros
[error]   private[collection] def nextPositivePowerOfTwo(target : Int): Int = 1 << -numberOfLeadingZeros(target - 1)
[error]                                                                             ^
[error] one error found
[error] (library/compile:compileIncremental) Compilation failed
[error] Total time: 18 s, completed Jan 21, 2017 8:55:02 AM
Build step 'Execute shell' marked build as failure
```

**Comment by l0rinc** [comment]
You're right, @retronym, thanks, fixed the missing import!

**Comment by SethTisue** [comment]
(We'll merge 2.11 onto 2.12 after 2.11.9 is out... annnnnnny day now.....)

**Comment by SethTisue** [comment]
The 2.11->2.12 merge is merged. Can you rebase?

**Comment by l0rinc** [comment]
@SethTisue, done! :)

**Review by SethTisue — APPROVED** [review]

**Comment by SethTisue** [comment]
thank you! including for your patience :-)

**Comment by l0rinc** [comment]
I wasn't in a hurry :)

**Pull request merged**